code/__DEFINES/milla.dm

```diff
@@ -0,0 +1,129 @@
+
+/* This comment bypasses grep checks */ /var/__milla
+
+/proc/__detect_milla()
+	if(world.system_type == UNIX)
+		// First check if it's built in the usual place.
+		if(fexists("./milla/target/i686-unknown-linux-gnu/release/libmilla.so"))
+			return __milla = "./milla/target/i686-unknown-linux-gnu/release/libmilla.so"
+		// Then check in the current directory.
+		if(fexists("./libmilla.so"))
+			return __milla = "./libmilla.so"
+		// And elsewhere.
+		return __milla = "libmilla.so"
+	else
+		// First check if it's built in the usual place.
+		if(fexists("./milla/target/i686-pc-windows-msvc/release/milla.dll"))
+			return __milla = "./milla/target/i686-pc-windows-msvc/release/milla.dll"
+		// Then check in the current directory.
+		if(fexists("./milla.dll"))
+			return __milla = "./milla.dll"
+		// And elsewhere.
+		return __milla = "milla.dll"
+
+#define MILLA (__milla || __detect_milla())
+
+#define MILLA_CALL(func, args...) call_ext(MILLA, "byond:[#func]_ffi")(args)
+
+#define REAL_SET_TILE_ATMOS(args...) MILLA_CALL(set_tile_atmos, args)
+
+/proc/set_tile_atmos(x, y, z, blocked_north, blocked_east, blocked_south, blocked_west, atmos_mode, external_temperature, oxygen, carbon_dioxide, nitrogen, toxins, sleeping_agent, agent_b, temperature, innate_heat_capacity)
+	if(!isnum(x) || !isnum(y) || !isnum(z))
+		CRASH("Called set_tile_atmos without specifying coordinates.")
+	if(!isnull(blocked_north) && !isnum(blocked_north))
+		CRASH("Called set_tile_atmos with non-number blocked_north.")
+	if(!isnull(blocked_east) && !isnum(blocked_east))
+		CRASH("Called set_tile_atmos with non-number blocked_east.")
+	if(!isnull(blocked_south) && !isnum(blocked_south))
+		CRASH("Called set_tile_atmos with non-number blocked_south.")
+	if(!isnull(blocked_west) && !isnum(blocked_west))
+		CRASH("Called set_tile_atmos with non-number blocked_west.")
+	if(!isnull(atmos_mode) && !isnum(atmos_mode))
+		CRASH("Called set_tile_atmos with non-number atmos_mode.")
+	if(!isnull(external_temperature) && !isnum(external_temperature))
+		CRASH("Called set_tile_atmos with non-number external_temperature.")
+	if(!isnull(oxygen) && !isnum(oxygen))
+		CRASH("Called set_tile_atmos with non-number oxygen.")
+	if(!isnull(carbon_dioxide) && !isnum(carbon_dioxide))
+		CRASH("Called set_tile_atmos with non-number carbon_dioxide.")
+	if(!isnull(nitrogen) && !isnum(nitrogen))
+		CRASH("Called set_tile_atmos with non-number nitrogen.")
+	if(!isnull(toxins) && !isnum(toxins))
+		CRASH("Called set_tile_atmos with non-number toxins.")
+	if(!isnull(sleeping_agent) && !isnum(sleeping_agent))
+		CRASH("Called set_tile_atmos with non-number sleeping_agent.")
+	if(!isnull(agent_b) && !isnum(agent_b))
+		CRASH("Called set_tile_atmos with non-number agent_b.")
+	if(!isnull(temperature) && !isnum(temperature))
+		CRASH("Called set_tile_atmos with non-number temperature.")
+	if(!isnull(innate_heat_capacity) && !isnum(innate_heat_capacity))
```

Comment:
Is this many null checks still performant?

Reply:
It seems to be fine in practice for set_tile_atmos, as it doesn't get called that many times.

Over about 100 seconds on a server that had seen lots of atmos activity:
![image](https://github.com/ParadiseSS13/Paradise/assets/98381/5f782259-91a6-4f74-93e0-888c8e3e39be)

get_tile_atmos is a bit more questionable, same profile run:
![image](https://github.com/ParadiseSS13/Paradise/assets/98381/2d1a906f-6416-4c99-9cd9-8ae96fdab805)

The others have much, much fewer calls, so the performance isn't significant there.

Here are the numbers from as much of server startup as I could capture:
![image](https://github.com/ParadiseSS13/Paradise/assets/98381/561000b9-5331-4113-a4e2-3736a3ff261a)
![image](https://github.com/ParadiseSS13/Paradise/assets/98381/10d7078b-8983-48a3-a8fb-e20586913fa6)

Seems to be the opposite situation there, get is OK but set is questionable.

Reply:
isnull is very performant on it's own and this actually called that often. It's mostly just on init and not much else.